contracts/tg4-stake/src/contract.rs

```diff
@@ -177,27 +211,58 @@ pub fn execute_unbond<Q: CustomQuery>(
         return Err(ContractError::InvalidDenom(denom));
     }
 
-    // reduce the sender's stake - aborting if insufficient
+    // Load stake first for comparison
+    let stake = STAKE
+        .may_load(deps.storage, &info.sender)?
+        .unwrap_or_default();
+    // Reduce the sender's stake - saturating if insufficient
     let new_stake = STAKE.update(deps.storage, &info.sender, |stake| -> StdResult<_> {
-        Ok(stake.unwrap_or_default().checked_sub(amount)?)
+        Ok(stake.unwrap_or_default().saturating_sub(amount))
     })?;
 
+    let mut res = Response::new()
+        .add_attribute("action", "unbond")
+        .add_attribute("amount", amount)
+        .add_attribute("denom", &denom)
+        .add_attribute("sender", &info.sender);
+
+    let mut new_vesting_stake = Uint128::zero();
+    if amount > stake {
+        // Reduce the sender's vesting stake - aborting if insufficient
+        let vesting_amount = amount - stake;
+        new_vesting_stake =
+            STAKE_VESTING.update(deps.storage, &info.sender, |stake| -> StdResult<_> {
+                Ok(stake.unwrap_or_default().checked_sub(vesting_amount)?)
+            })?;
+        // Undelegate (unstake from contract) to sender's vesting account
+        let msg = TgradeMsg::Undelegate {
+            funds: coin(vesting_amount.into(), cfg.denom.clone()),
+            recipient: info.sender.to_string(),
+        };
+        res = res
+            .add_message(msg)
+            .add_attribute("vesting_amount", vesting_amount);
+    }
+
+    // Create claim for unbonded liquid amount
     let completion = cfg.unbonding_period.after(&env.block);
     claims().create_claim(
         deps.storage,
         info.sender.clone(),
-        amount,
+        min(stake, amount),
         completion,
         env.block.height,
     )?;
+    res = res.add_attribute("completion_time", completion.time().nanos().to_string());
 
-    let mut res = Response::new()
-        .add_attribute("action", "unbond")
-        .add_attribute("amount", amount)
-        .add_attribute("denom", &denom)
-        .add_attribute("sender", &info.sender)
-        .add_attribute("completion_time", completion.time().nanos().to_string());
-    res.messages = update_membership(deps.storage, info.sender, new_stake, &cfg, env.block.height)?;
+    // Update membership messages
+    res = res.add_submessages(update_membership(
+        deps.storage,
+        info.sender,
+        new_stake + new_vesting_stake,
+        &cfg,
+        env.block.height,
+    )?);
```

Comment:
Are you sure those discarded attributes weren't useful? Just asking, I haven't used that contract "from the other side"

Reply:
They are above. Moved just for the updating logic.

Reply:
Right, I missed that.